File: iq/3/palindrome/solution_test.cc
// Copyright 2026 Samvel Khalatyan. All rights reserved.
//
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

//
// clang-format off-next-line
//go:build ignore

#include "iq/3/palindrome/solution.h"

#include <algorithm>
#include <cctype>
#include <iomanip>
#include <ostream>
#include <string>

#include "gtest/gtest.h"

namespace iq::palindrome {
namespace {

using ::testing::TestParamInfo;
using ::testing::TestWithParam;
using ::testing::ValuesIn;

constexpr char kCharUnderscore = '_';

struct IsTestParam {
  std::string name;
  std::string s;
  bool want;

  friend std::ostream& operator<<(std::ostream& os, const IsTestParam& tp) {
    return os << std::quoted(tp.name) << " s: " << std::quoted(tp.s);
  }
};

class IsTest : public TestWithParam<IsTestParam> {};

TEST_P(IsTest, pass) {
  const IsTestParam& tp = GetParam();

  const bool got = Is(tp.s);

  ASSERT_EQ(got, tp.want);
}

const IsTestParam kIsTestParams[]{
    {.name = "empty", .s = "", .want = true},
    {.name = "one letter", .s = "a", .want = true},
    {.name = "one digit", .s = "1", .want = true},
    {.name = "one non-alphanumeric letter", .s = ".", .want = true},
    {.name = "two letters", .s = "aa", .want = true},
    {.name = "two digits", .s = "11", .want = true},
    {.name = "two distinct digits", .s = "12", .want = false},
    {.name = "letter and digit", .s = "a1", .want = false},
    {.name = "letter and non-alphanumeric", .s = "a.", .want = true},
    {.name = "digit and non-alphanumeric", .s = "1.", .want = true},
    {.name = "non-alphanumeric and letter", .s = ".a", .want = true},
    {.name = "non-alphanumeric and digit", .s = ".1", .want = true},
};

INSTANTIATE_TEST_SUITE_P(IsTest, IsTest, ValuesIn(kIsTestParams),
                         [](const TestParamInfo<IsTestParam>& info) {
                           std::string name = info.param.name;
                           std::replace_if(
                               name.begin(), name.end(),
                               [](char c) { return !std::isalnum(c); },
                               kCharUnderscore);
                           return name;
                         });

}  // namespace
}  // namespace iq::palindrome

File: iq/3/palindrome/solution_test.go
// Copyright 2026 Samvel Khalatyan. All rights reserved.
//
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

package palindrome_test

import (
	"testing"

	"github.com/skhal/lab/iq/3/palindrome"
)

func TestIs(t *testing.T) {
	tests := []struct {
		name string
		str  string
		want bool
	}{
		{
			name: "empty",
			want: true,
		},
		{
			name: "one letter",
			str:  "a",
			want: true,
		},
		{
			name: "one digit",
			str:  "1",
			want: true,
		},
		{
			name: "non-alphanumic letter",
			str:  ".",
			want: true,
		},
		{
			name: "two letters",
			str:  "aa",
			want: true,
		},
		{
			name: "two digits",
			str:  "11",
			want: true,
		},
		{
			name: "two distinct letters",
			str:  "ab",
		},
		{
			name: "letter and digit",
			str:  "a1",
		},
		{
			name: "letter and non-alphanumeric",
			str:  "a.",
			want: true,
		},
		{
			name: "digit and non-alphanumeric",
			str:  "1.",
			want: true,
		},
		{
			name: "non-alphanumeric and letter",
			str:  ".a",
			want: true,
		},
		{
			name: "non-alphanumeric and digit",
			str:  ".1",
			want: true,
		},
	}
	for _, tc := range tests {
		tc := tc
		t.Run(tc.name, func(t *testing.T) {
			got := palindrome.Is(tc.str)
			if got != tc.want {
				t.Errorf("palindrome.Is(%q) = %v, want %v", tc.str, got, tc.want)
			}
		})
	}
}
